### ISSUE TITLE
autoUpdater.checkForUpdates() during an update emits an error on Windows

### ISSUE THREAD

**Issue by benjamin-wagon** [electron/electron#5097]
* Electron version: v0.37.3
* Operating system: Windows 10

Here's the stack trace that's included with the error event:

```
Command failed: 4294967295
System.AggregateException: One or more errors occurred.
---> System.Exception: Couldn't acquire lock, is another instance running
   at Squirrel.SingleGlobalInstance..ctor(String key, TimeSpan timeOut)
   at Squirrel.UpdateManager.<acquireUpdateLock>b__1e()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Squirrel.UpdateManager.<CheckForUpdate>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Squirrel.Update.Program.<Download>d__5d.MoveNext()
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at System.Threading.Tasks.Task`1.get_Result()
   at Squirrel.Update.Program.executeCommandLine(String[] args)
   at Squirrel.Update.Program.main(String[] args)
   at Squirrel.Update.Program.Main(String[] args)
---> (Inner Exception #0) System.Exception: Couldn't acquire lock, is another instance running
   at Squirrel.SingleGlobalInstance..ctor(String key, TimeSpan timeOut)
   at Squirrel.UpdateManager.<acquireUpdateLock>b__1e()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Squirrel.UpdateManager.<CheckForUpdate>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Squirrel.Update.Program.<Download>d__5d.MoveNext()<---
```

**Comment by kevinsawicki** [comment]
Yeah, I think the Windows implementation should probably guard against this case since the Mac version handles this gracefully, or at the very least mention this behavior in the docs.

**Comment by felixrieseberg** [comment]
@kevinsawicki: Should `checkForUpdates()` just do nothing - or emit an error?

**Comment by zcbenz** [comment]
This is actually a duplicate of #4306 and seems to be a by-design behavior of Squirrel, I don't know if it is possible to handle this on Electron's side, ignoring this error doesn't seem to be a solution.

**Comment by kevinsawicki** [comment]
My thoughts were that since Squirrel throws this error if two simultaneous update checks are performed then Electron could do one of the following:

*  Return immediately from `checkForUpdates` if an update check is already in progress since the events will still get emitted for the first check once if either completes, errors, or there is no update available
* Schedule the next check after the current check completes so checks are never simultaneously performed. So calling `checkForUpdates` twice would check for updates twice but serially instead of in parallel.

/cc @paulcbetts in case there might be Squirrel changes in progress or already completed that might alter this behavior.

**Comment by paulcbetts** [comment]
> Return immediately from checkForUpdates if an update check is already in progress since the events will still get emitted for the first check once if either completes, errors, or there is no update available

I'd try to convert checkForUpdates internally to use a Promise, and if >1 person calls it, they just get subscribed to that existing Promise

**Comment by felixrieseberg** [comment]
That sounds great - I'll get on that.
